### ISSUE TITLE
[go_router] choose in redirect method if `push` or `go`

### ISSUE THREAD

**Issue by dzziwny** [flutter/flutter#114131]
I would like to choose in `redirect` method if i want to clear a navigation history (as it is in `context.go`), even if i started navigation with `context.push`.

## Use case

On application start i have a few configuration pages that user setup once, and finally ends up in dashboard page.
Possible navigation flows (depend on what user already setup, and what skipped in the past):

```
'/' -> '/dashboard'
'/' -> '/language' -> '/dashboard'
'/' -> '/welcome'/ -> '/dashboard'
'/' -> '/welcome'/ -> '/language' -> '/dashboard'
```

In `language` page i want to have a back button in appbar and be able to pop page. When i ends up in `dashboard` i would like to have a clear navigation history.

## Proposal

1. There could be a `Redirection` class, with `clearHistory` parameter (`false` by default).

```
GoRoute(
      path: '/language,
      builder: (_, __) => const LanguagePage(),
      redirect: (_, __) {
          final skipToDashboard = ...;
          if (skipToDashboard) {
              return Redirection('/dashboard', clearHistory: true);
          }
          
          return null;
      }
    ),
```

2. Add `clearHistory` method to `GoRouter` class.

**Comment by exaby73** [comment]
Hello @dzziwny. Thank you for filing this request. Labelling the issue for visibility and insights from the team

**Comment by chunhtai** [comment]
you can have a top level route and use go to clear the history.

**Comment by dzziwny** [comment]
You mean, the top route over setup pages, like for ex. `/setup/language`? But if i go from `/setup/welcome` to `/setup/language` i need to use `push` anyway, right? The problem is that i don't want to decide from any setup page wether to use `push` or `go` - I am doing it now and have a separated service that resolves the next step and decide which method to use - it works but looks ugly and `redirect` lost any sense.

**Comment by hunterwilhelm** [comment]
Sometimes `redirect` does a `popUntil`. Demonstrated in #116164.

Could we add a way to choose between `push`, `go`, and `popUntil` too?

**Comment by Maczuga** [comment]
I have a similar request, that would require a redirect action to perform `push` instead of `go`.

I need to have routes under `/account` "guarded" with login check - so if user is not logged in - a sign in page shows up in root navigator (so over the `StackRoute` and `BottomNavigationBar`).
However I also need the `/user/login` to be regular route that can be called from anywhere and would behave the same.

Below is a code sample that parially works. By partially I mean that redirect works - it opens the `/user/login` route, it covers the screen, but there is no "back" arrow and the "back" action just closes the app.

<details>
  <summary>main.dart</summary>

```
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() => runApp(const MyApp());

// #region Router
final _rootNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'root');
final _shellNavigatorKey = GlobalKey<NavigatorState>(debugLabel: 'shell');

final _router = GoRouter(
  navigatorKey: _rootNavigatorKey,
  initialLocation: '/',
  redirect: (context, state) {
    if (state.location == "/account") return "/user/login";
    return null;
  },
  routes: [
    ShellRoute(
      navigatorKey: _shellNavigatorKey,
      builder: (context, state, child) {
        return ScaffoldWithNavBar(child: child);
      },
      routes: [
        GoRoute(
          path: "/",
          builder: (_, __) => const HomePage(),
        ),
        GoRoute(
          path: "/account",
          // redirect: (_, __) => "/user/login",

          pageBuilder: (BuildContext context, GoRouterState state) => MaterialPage<void>(
            key: state.pageKey,
            child: const AccountPage(),
          ),
        ),
      ],
    ),
    // I want this route to cover the entire screen (including bottom nav) - this works
    // However, I want to be able to get back from the redirect
    // So if user decides to NOT sign-in - he can go back to previous route
    // I assume that currently go_router uses .go, while I need push on redirect
    GoRoute(
      parentNavigatorKey: _rootNavigatorKey,
      path: "/user/login",
      builder: (_, __) => const LoginPage(),
    ),
  ],
);
// #endregion

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      title: 'Go Router Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      routerConfig: _router,
    );
  }
}

class ScaffoldWithNavBar extends StatelessWidget {
  final Widget child;

  const ScaffoldWithNavBar({
    required this.child,
    super.key,
  });

  static int _calculateSelectedIndex(BuildContext context) {
    final location = GoRouterState.of(context).location;
    if (location.startsWith('/home')) return 0;
    if (location.startsWith('/account')) return 1;

    return 0;
  }

  void _onItemTapped(int index, BuildContext context) {
    switch (index) {
      case 0:
        GoRouter.of(context).go('/home');
        break;
      case 1:
        GoRouter.of(context).go('/account');
        break;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: child,
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(Icons.home),
            label: 'Home',
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.person),
            label: 'Account',
          ),
        ],
        currentIndex: _calculateSelectedIndex(context),
        onTap: (int idx) => _onItemTapped(idx, context),
      ),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("HomePage")),
    );
  }
}

class AccountPage extends StatelessWidget {
  const AccountPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("AccountPage")),
    );
  }
}

class LoginPage extends StatelessWidget {
  const LoginPage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("LoginPage")),
      body: const Center(child: Text("This screen should cover the bottom nav bar, and have the 'back' arrow in AppBar.")),
    );
  }
}
```
</details>

**Comment by bipin-sakariya** [comment]
Hello @exaby73 ,
Any updates on this issue?

`final GoRouter router = GoRouter(
  routes: <RouteBase>[
    GoRoute(
      path: LoginMainPage.route,
      builder: (BuildContext context, GoRouterState state) {
        return LoginMainPage();
      },
      redirect: (BuildContext context, GoRouterState state) async {
        String appState = await AppFunctions.getAppState();
        if (state.fullpath == LoginMainPage.route) {
          if (appState == AppConstants.dashboard) {
            return DashboardMainPage.route;
          }
        }
        return null;
      },
      routes: <RouteBase>[
        GoRoute(
          path: 'register',
          builder: (BuildContext context, GoRouterState state) {
            return RegisterMainPage();
          },
        ),
        GoRoute(
          path: 'dashboard',
          builder: (BuildContext context, GoRouterState state) {
            return DashboardMainPage();
          },
          routes: <RouteBase>[
            GoRoute(
              path: 'profile',
              builder: (BuildContext context, GoRouterState state) {
                return const UpdateProfileMainPage();
              },
            ),
            GoRoute(
              path: 'questions',
              builder: (BuildContext context, GoRouterState state) {
                return QuestionMainPage();
              },
            ),
            GoRoute(
              path: 'result',
              builder: (BuildContext context, GoRouterState state) {
                return AssessmentMainPage();
              },
            )
          ],
        ),
      ],
    ),
  ],
);`

**Comment by mirkancal** [comment]
I'm trying to `push` page, instead of `go` with return.

My case is, in my redirect, I want to push the trial page(where you can start the trial) on top of the home page.

Based on my PurchasesBloc, when my offers are ready, I want to push the trial page wherever the user is.

When I just return /trial, it doesn't have any history and you can't go back with pushing.

I'm new in go_router and not sure if it's the correct place to return page or it's the correct issue, but it would fix my issue if I can push in redirect.

**Comment by noahge** [comment]
> Maczuga

same issue，any solve?

**Comment by Maczuga** [comment]
> > Maczuga
> 
> same issue，any solve?

Nope, ended up using stock navigator to create own routing solution cuz none of existing routing libraries were able to fill all my needs after all.

**Comment by patrikheinonen** [comment]
Any updates? redirect is broken on many levels see for example. https://github.com/flutter/flutter/issues/135359

**Comment by adonisRodxander** [comment]
Any update about how redirect using `push`?

**Comment by julgog** [comment]
I have a problem with go_router, if I am in
/home/details, where (details has a back button in the application bar) and within details a logout button, when the redirect function is activated.
I expect to go to /home, but what I get is the same location.
/home/details.
What is happening is this the expected behavior?

**Comment by SilentCatD** [comment]
> @patrikheinonen that issue seems to be a different problem.
> 
> I don't have a update for this issue, and no one is actively working on this. Something that would be tricky is that, let's say introduce a new redirect API, and that API would have function signature like.
> 
> ```dart
> void onRedirect(BuildContext context, GoRouterState state, GoRouter router) {
>   // directly call router.go or router.push, or any operations.
> }
> ```
> 
> It needs to be smart enough to know the operation doesn't redirect at this level and need to continue to call the redirect on the subroutes.
> 
> It seems like something that will be hard to implement. An alternative is implementing something like web's event.preventDefault or event.stopPropagate, but i don't have anything concrete yet.
> 
> I would like to have some suggestion or feedback on what the API should look like

How about make the redirect function return an object indicate next action instead of string?. For eg: we can have: `GoAction`, `PushAction`, `NoAction`. I think that is enough to handle this issue

**Comment by chunhtai** [comment]
> For eg: we can have: GoAction, PushAction, NoAction. I think that is enough to handle this issue

If we want to go this route, we can have
```dart
bool onRedirect(BuildContext context, GoRouterState state, GoRouter router) {
  // return true if redirect is handled
}
``` 
the push or go can be done directly in the callback.

**Comment by SilentCatD** [comment]
> > For eg: we can have: GoAction, PushAction, NoAction. I think that is enough to handle this issue
> 
> If we want to go this route, we can have
> ```dart
> bool onRedirect(BuildContext context, GoRouterState state, GoRouter router) {
>   // return true if redirect is handled
> }
> ``` 
> the push or go can be done directly in the callback.

If this is implemented, will it be a replacement for the current `redirect`, or a new function entirely?

**Comment by chunhtai** [comment]
Hi @SilentCatD 

I would replace it if we implement this. I don't think this add much boilerplate and can achieve everything `redirect` can do. Is there a concern if we do this?

**Comment by SilentCatD** [comment]
> Hi @SilentCatD 
> 
> I would replace it if we implement this. I don't think this add much boilerplate and can achieve everything `redirect` can do. Is there a concern if we do this?

Not at all, I think this will be really great. Thank you for your reply. This will solve a lot of use cases of in route guarding were it to be implemented, thanks.

**Comment by arthurbcd** [comment]
> > For eg: we can have: GoAction, PushAction, NoAction. I think that is enough to handle this issue
> 
> 
> 
> If we want to go this route, we can have
> 
> ```dart
> 
> bool onRedirect(BuildContext context, GoRouterState state, GoRouter router) {
> 
>   // return true if redirect is handled
> 
> }
> 
> ``` 
> 
> the push or go can be done directly in the callback.

Would it work like this?

```dart

 bool onRedirect(BuildContext context, GoRouterState state, GoRouter router) {

bool isLogged = ...;
if (!isLogged) router.go('/login');
return isLogged;

}

```

If so, then why we need the boolean? Shouldn't GoRouter be smart enough to detect if a .go or .push was called?

This way we can use this "wasCalled" condition to determine if redirect in handled.

**Comment by maherjaafar** [comment]
> > > For eg: we can have: GoAction, PushAction, NoAction. I think that is enough to handle this issue
> > 
> > 
> > If we want to go this route, we can have
> > ```dart
> > bool onRedirect(BuildContext context, GoRouterState state, GoRouter router) {
> >   // return true if redirect is handled
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > the push or go can be done directly in the callback.
> 
> If this is implemented, will it be a replacement for the current `redirect`, or a new function entirely?

It should be a new function.

**Comment by tp27933** [comment]
I am also staying on top of this feature and would like to know if there is any update.

**Comment by MiguelBelotto00** [comment]
Any news on this functionality or workaround?

**Comment by chunhtai** [comment]
sorry I was on vacation. This issue is still under my radar


> If so, then why we need the boolean? Shouldn't GoRouter be smart enough to detect if a .go or .push was called?

I think GoRouter can be smart enough to catch most cases by checking route match list changes, but there may be corner cases. e.g. the route history doesn't change after a series of push/pop. Even worse, some people may think this case should be considered redirect handled, while some people think otherwise.

We may even have cases where people want to mark it handled without doing any actual redirect.

**Comment by arthurbcd** [comment]
I believe the void redirect(context, state, router) is simpler to understand.

We would simply use the same API we already use to navigate. No extra booleans, "stopProgating" or other complicated API.

Why not start simple and just explain those edge cases?

Starting simple would already cover most cases:
- can't choose between push/go 
- can't set extra.
- simpler api

Otherwise this discussion will never ends.

**Comment by SilentCatD** [comment]
Can we also consider the importance of seeing the requested route's navigation method (push, go, ....) 

For example, the requested route must be navigated by the go method. But the route guard (login) will be pushed.

Knowing the navigation methods will allow us to continue the previously requested navigation scheme.

**Comment by chunhtai** [comment]
I wrote a design doc. Let me know what you think

http://flutter.dev/go/go-router-redirect

**Comment by maherjaafar** [comment]
@chunhtai It looks good. 

I would choose `2.1. Alternative 1: imperative builder API [currently preferred]` 
because I don't like `SplashBuilder.of(context).builder = splashBuilder;` in 2.2

**Comment by arthurbcd** [comment]
I love the idea of having a list of Guards! The addition of currentState and nextState is also a great feature.

The splash seems something too manual, can't we have a `loadingBuilder` in the `GoRoute` to show while all Guards are resolving? 

About onEnter/onExit:
The current onExit api has a `canExit` boolean that we need to return. Since this is already implemented, maybe we can keep the same idea of it by also using a `canEnter` boolean in replacement of the `Resolution.next`.

I agree with you, `onExit` and `onEnter` could be unified like you suggested at the end (as long as they have the same api). Fewer breaking changes by doing this altogether.

One doubt. What happens if you return `Resolution.stop` without navigating at all? Will you stay in the previous route or is it an error?

Great work as always @chunhtai. Thanks.

**Comment by feinstein** [comment]
@chunhtai will this solution allow us to `go` and `push` from the `redirect`?

In my case I have a deeplink that goes to a page that's always pushed. But depending on the page the user is now, I don't want to push it, I need to first `go` to a particular page (a `StatefulShellRoute`), and then `push` (so the new page covers the bottom navigation bar inside the `StatefulShellRoute` as well).

I imagine I could craft some logic inside the `go`, to add some extra `queryParameter` and then redirect again with a `push`, but this is more cumbersome.

**Comment by MostafaSolimanMO** [comment]
Any updates on this one?

**Comment by cedvdb** [comment]
@chunhtai regarding your design document, have you converged towards the final design ? It seems like it would solve a lot of issues mentioned.

Are there plans to implement this soon or is this something that's not planned ?

**Comment by loveoverflowdev** [comment]
I have the same request

**Comment by christo-caddayn** [comment]
I did a workaround. I don't know wheter it suits other usecases. 

FutureOr<String?> redirect(BuildContext context, GoRouterState state) async {
  if (!isAppLatest) {   // REDIRECTION CONDITION
    final p = 'my-path-to/screen';
    String data = 'sample data';

    Future.delayed(Duration(milliseconds: 100),
        () => GoRouter.of(context).go(p, extra: data)); // ADDING A FUTURE CALL WHICH WILL BE CALLED AFTER REDIRECTION
    return p;
  }
}

I know this is not the best way. Just posting here so that anyone looking for workarounds can try this.

**Comment by abdullahrs** [comment]
This feature is crucial for complex deep-linking scenarios. Simply redirecting to the given location causes the loss of the existing navigation stack.

**Comment by chunhtai** [comment]
This pr should support this feature https://github.com/flutter/packages/pull/8339
